src/System.Runtime.CompilerServices.Unsafe/src/System.Runtime.CompilerServices.Unsafe.ilproj

```diff
@@ -0,0 +1,94 @@
+  </PropertyGroup>
+
+  <ItemGroup>
+    <!-- mscorlib is passed in as an explicit reference from C# targets but not via the IL SDK. -->
+    <Reference Include="$(CoreAssembly)" Condition="'$(TargetFramework)' != 'netstandard2.0'" />
+  </ItemGroup>
+
+  <Target Name="GenerateVersionFile"
+          DependsOnTargets="GetAssemblyVersion;ResolveReferences"
+          Inputs="$(MSBuildAllProjects)"
+          Outputs="'$(VersionFilePath)">
+   <PropertyGroup>
+      <IncludePath>$([MSBuild]::NormalizeDirectory('$(IntermediateOutputPath)', 'version'))</IncludePath>
+      <IncludePathTrimmed>$(IncludePath.TrimEnd('\'))</IncludePathTrimmed>
+      <IlasmFlags>$(IlasmFlags) -INCLUDE="$(IncludePathTrimmed)"</IlasmFlags>
+      <VersionFilePath Condition="'$(VersionFilePath)' == ''">$([MSBuild]::NormalizePath('$(IncludePath)', 'version.h'))</VersionFilePath>
+      <_AssemblyVersion>$(AssemblyVersion.Replace('.', ':'))</_AssemblyVersion>
+      <_coreAssemblyName Condition="'%(ReferencePath.FileName)' == '$(CoreAssembly)'">%(ReferencePath.FusionName)</_coreAssemblyName>
+      <_assemblyNamePattern><![CDATA[[^,]+, Version=(?<v1>[0-9]+)\.(?<v2>[0-9]+)\.(?<v3>[0-9]+)\.(?<v4>[0-9]+), .*PublicKeyToken=(?<p1>[0-9a-f]{2})(?<p2>[0-9a-f]{2})(?<p3>[0-9a-f]{2})(?<p4>[0-9a-f]{2})(?<p5>[0-9a-f]{2})(?<p6>[0-9a-f]{2})(?<p7>[0-9a-f]{2})(?<p8>[0-9a-f]{2})]]></_assemblyNamePattern>
+      <_coreAssemblyVersion>$([System.Text.RegularExpressions.Regex]::Replace(
+        $(_coreAssemblyName),
+        $(_assemblyNamePattern),
+        '${v1}:${v2}:${v3}:${v4}'))</_coreAssemblyVersion>
+      <_coreAssemblyPublicKeyToken>$([System.Text.RegularExpressions.Regex]::Replace(
+        $(_coreAssemblyName),
+        $(_assemblyNamePattern),
+        '${p1} ${p2} ${p3} ${p4} ${p5} ${p6} ${p7} ${p8}').ToUpperInvariant())</_coreAssemblyPublicKeyToken>
+      <_VersionFileContents>
+<![CDATA[
+#define CORE_ASSEMBLY "$(CoreAssembly)"
+#define ASSEMBLY_VERSION "$(_AssemblyVersion)"
+#define CORE_ASSEMBLY_VERSION "$(_CoreAssemblyVersion)"
+#define FILE_VERSION "{string('$(FileVersion)')}"
+#define INFORMATIONAL_VERSION "{string('$(InformationalVersion)')}"
+$(ExtraMacros)
+// Metadata version: v4.0.30319
+.assembly extern CORE_ASSEMBLY
+{
+  .publickeytoken = ($(_coreAssemblyPublicKeyToken) )
+  .ver CORE_ASSEMBLY_VERSION
+}
+ ]]>
+      </_VersionFileContents>
+    </PropertyGroup>
+
+    <WriteLinesToFile
+      File="$(VersionFilePath)"
+      Lines="$(_VersionFileContents)"
+      Overwrite="true"
+      WriteOnlyWhenDifferent="true" />
+
+    <ItemGroup>
+      <FileWrites Include="$(VersionFilePath)" />
+    </ItemGroup>
+  </Target>
+
+  <!-- Decompile the ILResourceReference to get native resources. -->
+  <Target Name="SetILResourceReference"
+          BeforeTargets="DisassembleIlasmResourceFile"
+          Condition="'@(ResolvedMatchingContract)' != ''">
+    <ItemGroup>
+      <ILResourceReference Include="@(ResolvedMatchingContract)" />
```

Comment:
Is this working correctly?  I noticed that the commandline to ILASM didn't have resource.  In this repo we don't build a contract assembly, so you might want to use some other CSProj to produce a binary that can be used for grabbing native version resources.

Reply:
I don't really understand this. Can you explain what you mean by using another csproj?

Reply:
Ilasm doesn't have a way for defining PE version resources in IL.  Instead what you need to do is pass in a resource object file.

Rather than having a scheme for authoring native resources as `.rc` and flowing all the right branding info into that ILProj uses a trick.  It produces resources for the ILProj by extracting them from some other assembly built with CSC.  https://github.com/dotnet/runtime/blob/57f870f909dbfad35142e5aaa6e681464de4f439/src/coreclr/.nuget/Microsoft.NET.Sdk.IL/Microsoft.NET.Sdk.IL.targets.template#L71-L111

So you just need an assembly that's versioned appropriately and that needs to be passed in as the `ILResourceReference` item.  You'll be able to see if you've gotten it correct by examining the version properties of the file (for instance in Windows Explorer).

Setting the native version resources is important because this is used by installers for versioning.

Related: https://github.com/dotnet/runtime/issues/11412

Reply:
> So you just need an assembly that's versioned appropriately 

I added the missing versioning properties in the ilproj. But if I look at the properties of the 

![image](https://github.com/user-attachments/assets/d081830b-8a33-43c3-8c45-685d57a4a564)

According to the binlog, that target is not getting evaluated due to the condition not being met:

> Target "SetILResourceReference" skipped, due to false condition; ('@(ResolvedMatchingContract)' != '') was evaluated as ('' != '').

I'm now trying to understand how or where we gave a value to `ResolvedMatchingContract` in release/6.0.

Reply:
That's coming from the reference assembly resolution.  We won't have a reference assemblies here, so we can't use that to feed into the versioning.  Instead we just need some other CSProj to feed in for version information.  We could create a dummy one (probably the cleanest) or use an existing one (hacky, but doesn't add projects).  In either case we'd need to be careful to not include that reference in the package (as dependency or file).